core/trino-main/src/main/java/io/trino/sql/gen/columnar/BloomColumnarFilter.java

```diff
@@ -0,0 +1,235 @@
+import io.trino.spi.type.Type;
+import io.trino.sql.planner.BloomFilterWithRange;
+import org.objectweb.asm.MethodTooLargeException;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import static io.airlift.bytecode.Access.FINAL;
+import static io.airlift.bytecode.Access.PRIVATE;
+import static io.airlift.bytecode.Access.PUBLIC;
+import static io.airlift.bytecode.Access.a;
+import static io.airlift.bytecode.Parameter.arg;
+import static io.airlift.bytecode.ParameterizedType.type;
+import static io.airlift.bytecode.expression.BytecodeExpressions.add;
+import static io.airlift.bytecode.expression.BytecodeExpressions.constantInt;
+import static io.airlift.bytecode.expression.BytecodeExpressions.lessThan;
+import static io.trino.cache.CacheUtils.uncheckedCacheGet;
+import static io.trino.cache.SafeCaches.buildNonEvictableCache;
+import static io.trino.spi.StandardErrorCode.COMPILER_ERROR;
+import static io.trino.spi.StandardErrorCode.QUERY_EXCEEDED_COMPILER_LIMIT;
+import static io.trino.sql.gen.columnar.ColumnarFilterCompiler.updateOutputPositions;
+import static io.trino.util.CompilerUtils.defineClass;
+import static io.trino.util.CompilerUtils.makeClassName;
+import static io.trino.util.Reflection.constructorMethodHandle;
+
+public class BloomColumnarFilter
+{
+    // Generate a BloomColumnarFilter class per Type to avoid mega-morphic call site when reading a position from input block
+    private static final NonEvictableCache<Type, MethodHandle> COLUMNAR_BLOOM_FILTER_CACHE = buildNonEvictableCache(
+            CacheBuilder.newBuilder()
+                    .maximumSize(100)
+                    .expireAfterWrite(2, TimeUnit.HOURS));
+
+    private BloomColumnarFilter() {}
+
+    public static Supplier<FilterEvaluator> createBloomFilterEvaluator(BloomFilterWithRange bloomFilterWithRange, int inputChannel)
+    {
+        return () -> new ColumnarFilterEvaluator(
+                new DictionaryAwareColumnarFilter(
+                        createColumnarBloomFilter(bloomFilterWithRange.type(), inputChannel, bloomFilterWithRange).get()));
+    }
+
+    private static Supplier<ColumnarFilter> createColumnarBloomFilter(Type type, int inputChannel, BloomFilterWithRange bloomFilterWithRange)
+    {
+        MethodHandle filterConstructor = uncheckedCacheGet(
+                COLUMNAR_BLOOM_FILTER_CACHE,
+                type,
+                () -> generateColumnarBloomFilterClass(type));
+
+        return () -> {
+            try {
+                InputChannels inputChannels = new InputChannels(ImmutableList.of(inputChannel), ImmutableList.of(inputChannel));
+                return (ColumnarFilter) filterConstructor.invoke(bloomFilterWithRange, inputChannels);
+            }
+            catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        };
+    }
+
+    private static MethodHandle generateColumnarBloomFilterClass(Type type)
```

Comment:
Do we need to generate class? Is class load isolation not enough?

Reply:
Classloader isolation would also work, I used code generation to be consistent with what we do in the rest of the columnar expression compiler code.